### PR TITLE
Fix Deformable Convolutional Op Support #179

### PR DESCRIPTION
**Pull request by smallsunsun1**
Add cpu and gpu implementation for Deformable Conv2d and DeformablePSROIAlign according to https://arxiv.org/abs/1811.11168, EquiConv according to https://arxiv.org/pdf/1903.08094.pdf. 
The original https://github.com/tensorflow/addons/pull/1056 has closed because of use wrong email-address in commit history.

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
@googlebot I consent

**Comment by gabrieldemarmiesse** [comment]
@googlebot I consent

**Comment by gabrieldemarmiesse** [comment]
We used black to format some files in the master branch. To avoid future conflicts, I updated your pull request. Please do `git pull` before working on it again locally.

**Comment by vlordier** [comment]
@smallsunsun1 when swapping `Conv2D` with `EquiConv`, `kernel_size` and `strides` need to be changed from `int` to `tuple`. 
It might be helpful to align `Conv2D` and `EquiConv` prototypes

**Comment by smallsunsun1** [comment]
> @smallsunsun1 when swapping `Conv2D` with `EquiConv`, `kernel_size` and `strides` need to be changed from `int` to `tuple`.
> It might be helpful to align `Conv2D` and `EquiConv` prototypes as in [Conv2D prototype](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2D)
> 
> where `kernel_size` or `strides` is An integer or tuple/list of 2 integers

Thanks for you suggestion, I will align `Conv2D` and `EquiConv` prototypes in later commit 0.0 . Now EquiConv support dynamic input and change all the numpy operation into TF operation.

**Comment by smallsunsun1** [comment]
I think I have already finish all the work. Can any one review this PR and give some comments. @seanpmorgan @Squadrick

**Comment by Squadrick** [comment]
@smallsunsun1 Can you pull from `master` and update this PR?

Also, please take a look at #1328, we've changed our testing procedure. 

Sorry about the delay, I'll review the PR soon.

**Comment by smallsunsun1** [comment]
> @smallsunsun1 Can you pull from `master` and update this PR?
> 
> Also, please take a look at #1328, we've changed our testing procedure.
> 
> Sorry about the delay, I'll review the PR soon.

I merge the upstream master into this, but now have some many errors, maybe will take some time to fix... and my cla is also failed and i can't fix it .....

**Comment by Squadrick** [comment]
@smallsunsun1 See [this](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#development-tips) for auto formatting your code.

**Comment by smallsunsun1** [comment]
@Squadrick  All code has be formatted, but the CLA problem still exists.  can you fix it on your side.

**Comment by mdv3101** [comment]
@Squadrick can you please merge the code. I need to integrate deformable convolution with my own model. Waiting for it to get integrated with tensorflow.

**Comment by seanpmorgan** [comment]
@googlebot I consent.

**Comment by Squadrick** [comment]
@mdv3101 This is quite a massive PR, it'll take me a while to review and test its correctness. If you want it immediately, you can clone @smallsunsun1's repo and test it out yourself (in case there's any issues you can report back). Also, any help with reviews would be much appreciated too.

**Comment by lzx551402** [comment]
@smallsunsun1 Thank you so much for making this PR, which would be a huge contribution to this community! Do you think you can also include [DeformConvV2](https://arxiv.org/abs/1811.11168) in this implementation? Like, to optionally allow for an input that assigns the scaling factor for each grid point?

**Comment by tuanhanhdmprof** [comment]
> @mdv3101 This is quite a massive PR, it'll take me a while to review and test its correctness. If you want it immediately, you can clone @smallsunsun1's repo and test it out yourself (in case there's any issues you can report back). Also, any help with reviews would be much appreciated too.

Hi bro,
How is the status of this PR?  And Which one is @smallsunsun1's repo for?

**Comment by bhack** [comment]
> This is quite a massive PR, it'll take me a while to review and test its correctness. If you want it immediately

It is `+4,614 −0 ` probably you need to split this in multiple PRs

**Comment by ujjwal-ai** [comment]
Any progress on this ?

**Comment by mohamedmindee** [comment]
It would be nice to use deformable convolutions with tensorflow. When do you think it will be merged ?

**Comment by WindQAQ** [comment]
This PR is quite stale. I would recommend that we can split kernels/layers and example into different PR to help us review faster. Also, there are many code segments that can be directly reused from core TF such as `CheckFormatConstraintsOnShape`, `DimensionsFromShape`, `FastBoundsCheck` e.t.c. Please update them accordingly. If @smallsunsun1 doesn't mind, I can help with this.

**Comment by smallsunsun1** [comment]
> This PR is quite stale. I would recommend that we can split kernels/layers and example into different PR to help us review faster. Also, there are many code segments that can be directly reused from core TF such as `CheckFormatConstraintsOnShape`, `DimensionsFromShape`, `FastBoundsCheck` e.t.c. Please update them accordingly. If @smallsunsun1 doesn't mind, I can help with this.

It is appreciate if you can help deal with these. I don't have much time to update this PR now.

**Comment by WindQAQ** [comment]
> > This PR is quite stale. I would recommend that we can split kernels/layers and example into different PR to help us review faster. Also, there are many code segments that can be directly reused from core TF such as `CheckFormatConstraintsOnShape`, `DimensionsFromShape`, `FastBoundsCheck` e.t.c. Please update them accordingly. If @smallsunsun1 doesn't mind, I can help with this.
> 
> It is appreciate if you can help deal with these. I don't have much time to update this PR now.

Sure, then I'll open a new PR and ping you.

**Comment by oubathAI** [comment]
Hi @smallsunsun1 

Thanks for the nice PR.
I merged your PR locally and I was able to run DeformableConv2D without any major problems.

However, when I test DeformableConv2D on Scaled MNIST, the results are very variable depending on the im2col parameter.
When im2col = 1 and batch_size=32 the results are rather good (accuracy > 95% in 5 epochs) ; but when im2col = 32 with batch_size=32, the results are very bad (accuracy < 50% in 5 epochs). 
Is this drop in accuracy normal? If so, can you explain me the reasons?

I also found this implementation very similar to yours: https://github.com/BIGKnight/deformable_conv2d_v2_tensorflow

In this repository, the author mentions that it is necessary to make a swapaxis if we want to use im2col > 1. In the file deformable_conv2d_op/deformable_conv2d.cc line 333, you can even see which line the author has planned to do it (https://github.com/BIGKnight/deformable_conv2d_v2_tensorflow/blob/e0491b3bae08561a6a932000e535759c605b0837/deformable_conv2d_op/deformable_conv2d.cc#L333); however I don't see any trace of swapaxis in your code. 
Can you explain to me why you removed the swapaxis? Isn't that the reason for bad learning when im2col > 1 ?

Thank you in advance for your answer

**Comment by smallsunsun1** [comment]
When im2col &gt; 1, current implementation is wrong.
The c++ part of DeformableConv2D is largely based on this repo(https://github.com/BIGKnight/deformable_conv2d_v2_tensorflow，DeformablePSRoIAlign code&nbsp; mainly based on official Pytorch version. I think the Parameter im2col do nothing with the speed of kernel execution, when im2col not equal to 1, current implemention does not work. at first I want to add assertion in python code to ensure and tell users im2col always to be 1, but finally I forgot it.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "tensorflow/addons"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 凌晨0:38
收件人:&nbsp;"tensorflow/addons"<addons@noreply.github.com&gt;;
抄送:&nbsp;"sss"<975759105@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [tensorflow/addons] Fix Deformable Convolutional Op Support #179 (#1129)





 
Hi @smallsunsun1
 
Thanks for the nice PR.
 I merged your PR locally and I was able to run DeformableConv2D without any major problems.
 
However, when I test DeformableConv2D on Scaled MNIST, the results are very variable depending on the im2col parameter.
 When im2col = 1 and batch_size=32 the results are rather good (accuracy &gt; 95% in 5 epochs) ; but when im2col = 32 with batch_size=32, the results are very bad (accuracy < 50% in 5 epochs).
 Is this drop in accuracy normal? If so, can you explain me the reasons?
 
I also found this implementation very similar to yours: https://github.com/BIGKnight/deformable_conv2d_v2_tensorflow
 
In this repository, the author mentions that it is necessary to make a swapaxis if we want to use im2col &gt; 1. In the file deformable_conv2d_op/deformable_conv2d.cc line 333, you can even see which line the author has planned to do it (https://github.com/BIGKnight/deformable_conv2d_v2_tensorflow/blob/e0491b3bae08561a6a932000e535759c605b0837/deformable_conv2d_op/deformable_conv2d.cc#L333); however I don't see any trace of swapaxis in your code.
 Can you explain to me why you removed the swapaxis? Isn't that the reason for bad learning when im2col &gt; 1 ?
 
Thank you in advance for your answer

**Comment by oubathAI** [comment]
Thanks for your quick reply.

> when im2col not equal to 1, current implemention does not work

This corresponds to my observation; thanks for your confirmation. 

> The c++ part of DeformableConv2D is largely based on this repo (https://github.com/BIGKnight/deformable_conv2d_v2_tensorflow

I thought so. There's a lot of identical code between your two repositories. But why didn't you mention that anywhere? I didn't see any reference to his work in your code.

> I think the Parameter im2col do nothing with the speed of kernel execution,

I'm not sure about that statement, what do you base it on?
The code you rely on is itself an adaptation of the code repository [Deformable-ConvNets](https://github.com/msracver/Deformable-ConvNets). 
It is clearly stated in the [FAQ](https://github.com/msracver/Deformable-ConvNets/blob/master/DCNv2_op/README.md) of the original code that im2col can enhance performance; I quote "_The updated operator is significantly faster than the existing one when the image batch size is large_". 
This seems logical when you look at the code; indeed in the PR you submitted, if you look at the file [deformable_conv_op.cc line 1017](https://github.com/tensorflow/addons/blob/72c19cfb6c6b0a6ae1bbe9fa736811cfe02a3b9a/tensorflow_addons/custom_ops/layers/cc/kernels/deformable_conv_op.cc#L1017) you can see that the loop `for (int32_t n = 0; n < num_ / im2col_step_; ++n)` depends on im2col; the greater im2col will be, the less loops will be; therefore the calculation will be faster. 

From what I've seen from BIGKnight's code but also from the original code, `swapaxis` is necessary to perform the operations with im2col (even if I didn't understand why). I guess the reason you didn't include `swapaxis` in your code is because you couldn't port it?

Thanks in advance for your answer

**Comment by smallsunsun1** [comment]
My previous conclusion is wrong. I read the code now, confirmed that there will be less times to call gpu and cpu kernel function if im2col become larger. maybe at first implentation, I just guarantee the cpu and gpu implementation are right, to be honest, this repo is my first time to write PR and so large cpp code, there always to have some problem..... you can update the code to make im2col work both in cpu and gpu mode.&nbsp;

---Original---

**Comment by axeldavy** [comment]
Hi, I tried to run the PR on tensorflow 2.3 and had the following errors:
. Some ' #include "tensorflow/core/framework/kernel_shape_util.h" ' were missing causing compilation failures
. When using mixed_float16, the layer triggers an error saying it needs float32 or float64. If passing dtype='float32', it fails saying offset is float16 while input is float32. The fix is to pass to the conv2D producing 'offset' the argument dtype.
. get_config is missing 'kernel_initializer', etc

I thought I'd let you know so it can be fixed before the PR is merged :-).

**Comment by alexattia** [comment]
Looking forward to it! Do you any news about the progress of this PR ?

**Comment by qin2294096** [comment]
I have try BIGKnight's code in my project, replacing three `conv2d` layers by `deformable_conv2d `in `center-net-resnet18`.
And I saw:
1. the training time is about twice as much as before
2. backward is really slow
3. `im2col_step` has no effect on training speed

environment: tf1.15, cuda10, cudnn7, gtx1080ti
`batchsize`=16, `im2col_step`=1
image

`batchsize`=16, `im2col_step`=16
image

No `deformable_conv2d`
image

By the way, in my project, use `deformable_conv2d` with `im2col_step`=1 brings a good performance (haven't try `im2col_step`=16). If the training speed becomes faster, it will be a very good job!

**Comment by bhack** [comment]
I don't think we have the bandwidth to maintain this custom ops. If you are still interested please contribute the python layer  to https://github.com/keras-team/keras-cv.
If you need more native ops please contribute these to TensorFlow.
